drivers/flash/flash_bflb.c

```diff
@@ -0,0 +1,1564 @@
+
+	/* reenable burst read */
+	if ((data->flash_cfg.io_mode & 0x10) != 0) {
+		if ((data->flash_cfg.io_mode & 0xf) == 2 || (data->flash_cfg.io_mode & 0xf) == 4) {
+			ret = flash_bflb_flash_enable_burst(data);
+			if (ret != 0) {
+				return ret;
+			}
+		}
+	}
+
+	ret = flash_bflb_xip_init(data);
+	if (ret != 0) {
+		return ret;
+	}
+
+	return 0;
+}
+
+#if defined(CONFIG_SOC_FLASH_BFLB_DIRECT_ACCESS)
+
+static int flash_bflb_read_sahb_do(struct flash_bflb_data *data, off_t address, void *buffer,
+				   size_t length)
+{
+	int ret;
+	struct bflb_flash_command read_cmd = {0};
+	size_t i, cur_len;
+
+	read_cmd.spi_mode = data->flash_cfg.io_mode & 0xf;
+
+	switch (data->flash_cfg.io_mode & 0xf) {
+	default:
+	case 0:
+		read_cmd.cmd_buf[0] = data->flash_cfg.fast_read_cmd << 24;
+		read_cmd.dummy_clks = data->flash_cfg.fr_dmy_clk;
+	break;
+	case 1:
+		read_cmd.cmd_buf[0] = data->flash_cfg.fast_read_do_cmd << 24;
+		read_cmd.dummy_clks = data->flash_cfg.fr_do_dmy_clk;
+	break;
+	case 2:
+		read_cmd.cmd_buf[0] = data->flash_cfg.fast_read_qo_cmd << 24;
+		read_cmd.dummy_clks = data->flash_cfg.fr_qo_dmy_clk;
+	break;
+	case 3:
+		read_cmd.cmd_buf[0] = data->flash_cfg.fast_read_dio_cmd << 24;
+		read_cmd.dummy_clks = data->flash_cfg.fr_dio_dmy_clk;
+	break;
+	case 4:
+		read_cmd.cmd_buf[0] = data->flash_cfg.fast_read_qio_cmd << 24;
+		read_cmd.dummy_clks = data->flash_cfg.fr_qio_dmy_clk;
+	break;
+	}
+	read_cmd.addr_size = 3;
+
+	/* continuous read for qo and qio */
+	if ((data->flash_cfg.io_mode & 0xf) == 2 || (data->flash_cfg.io_mode & 0xf) == 4) {
+
+		if ((data->flash_cfg.c_read_support & 0x02) == 0) {
+			if ((data->flash_cfg.c_read_support & 0x01) == 0) {
+				/* "Not support cont read,but we still need set read mode(winbond
+				 * 80dv)"
+				 */
+				read_cmd.cmd_buf[1] = data->flash_cfg.c_read_mode << 24;
+			} else {
+				read_cmd.cmd_buf[1] =  data->flash_cfg.c_read_mode << 24;
+			}
+			read_cmd.addr_size = 4;
+		}
+	}
+
+	i = 0;
+	while (i < length) {
+
+		cur_len = data->flash_cfg.page_size - ((address + i) % data->flash_cfg.page_size);
+
+		if (cur_len > length - i) {
+			cur_len = length - i;
+		}
+
+		read_cmd.cmd_buf[0] &= ~0xFFFFFF;
+		read_cmd.cmd_buf[0] |= (address + i);
+		read_cmd.nb_data = cur_len;
+
+		ret = flash_bflb_send_command(data, &read_cmd);
+		if (ret != 0) {
+			return ret;
+		}
+
+		flash_bflb_xip_memcpy((uint8_t *)SF_CTRL_BUF_BASE, (uint8_t *)(buffer) + i,
+				      cur_len);
+
+		i += cur_len;
+
+		if (flash_bflb_busy_wait(data)) {
+			return -EBUSY;
+		}
+
+		if (flash_bflb_flash_busy_wait(data)) {
+			return -EBUSY;
+		}
+	}
+
+	return 0;
+}
+
+/* copies flash data using direct access */
+static int flash_bflb_read(const struct device *dev, off_t address, void *buffer, size_t length)
+{
+	struct flash_bflb_data *data = dev->data;
+	unsigned int	locker;
+	int ret;
+
+	ret = flash_bflb_is_valid_range(address, length);
+	if (ret != 0) {
+		return ret;
+	}
+
+	if (length == 0) {
+		return 0;
+	}
```

Comment:
In all cases, lack of action should not generate an error. This means the length check should go above range check.

Reply:
done